Fixing bugs with cards

Cattharine opened the pull request:
+ userstory cards shouldn't be movable after moving one userstory card in the beginning of the game
 + one shouldn't be able to add any card to one place twice

Cattharine committed:
Fixing bugs with cards
 + userstory cards shouldn't be movable after moving one userstory card in the beginning of the game
 + one shouldn't be able to add any card to one place twice

krutovsky-danya commented on game/game.py:
```diff
@@ -222,32 +223,31 @@ def _is_ready_to_spawn_tech_debt(self):
-    def _on_userstory_card_dropped(self, card, is_on_left: bool):
+    def _on_userstory_card_dropped(self, card: UserStoryCard, is_on_left: bool):
         if is_on_left:
+            if not card.is_in_release:
```
Это же and

krutovsky-danya commented on game/game.py:
```diff
@@ -222,32 +223,31 @@ def _is_ready_to_spawn_tech_debt(self):
+                return
             self.userstories.on_stories_card_dropped(card)
         else:
+            if card.is_in_release:
```
И elif-ами можно пользоваться)

krutovsky-danya commented on test/test_game_functions.py:
```diff
@@ -125,6 +127,75 @@ def test_can_start_blank_sprint(self):
+        self.assertEqual(len_us_release_before, len(self.game.userstories.release))
+
+    def test_cannot_move_one_card_to_sprint_twice(self):
+        card = Card()
```
Не равнозначные названия, тогда нужно card1

Cattharine committed:
Edit commented

krutovsky-danya commented on game/game.py:
```diff
@@ -237,15 +237,12 @@ def buy_room(self, room_num):  # !
-            if card.is_in_release:
-                return
+            card.is_in_release = False
+        elif not card.is_in_release:
```
Тут правда теперь условия не то. Раньше было `if not is_on_left and not card.is_in_release`, а теперь через или

Cattharine committed:
Fix condition and starting sprint permission

Cattharine committed:
Fixing environment state

There is a problem with `context.current_sprint_hours` being updated only after calling function that starts sprint in game or that moves a userstory card. Because of that environment state contained unexpected values for current sprint hours most of the time

krutovsky-danya commented on game/backlog/backlog.py:
```diff
@@ -18,7 +18,7 @@ def can_start_sprint(self):
         if hours_to_sum > self.get_max_hours():
             return False
-        return int(hours_to_sum) != 0 or int(self.context.customers) != 0
+        return hours_to_sum != 0 or self.context.customers > 0 or self.context.customers < 0
```
Все хорошо, только это выгладит странно
Можно заменить на abs(...) > 0 или оставить != 0

Cattharine committed:
Add required changes

krutovsky-danya commented on environment/environment.py:
```diff
@@ -77,7 +77,7 @@ def _get_state(self, in_tensor=False):
             context.credit / 10 ** 5,
             context.available_developers_count,
             context.current_rooms_counter,
-            context.current_sprint_hours,
```
А давай еще и эту строку оставим)

Cattharine replied:
С первым (про перенос карточки) согласна, такое изменение я вносила, но оно потерялось при переносе. Внесла изменение
Для второго (оставлять строку с `context.current_sprint_hours`) я пока что-то не вижу причин. Пояснишь?

krutovsky-danya replied:
Просто из соображения больше информации - лучше
Если дать ей оба кусочка данных, она легко сможет понять, что роботы будут работать и потратят деньги
Раньше сети приходилось самой складывать часы, чтобы не получить штраф за действие, а сейчас не придется, тем самым ускорится обучение
Если дать ей оба кусочка данных, она легко сможет понять, что роботы будут работать и потратят деньги
(Понял, что из-за лишнего добавления карточек в сеть, она не может посчитать сумму, нужно будет либо переделать, либо понять, что это не мешает)

krutovsky-danya reviewed (changes requested):
Сорри, еще докопаюсь:
В наградах за действия нет проверки, что карточку можно двигать и появляется положительная награда за не валидное действие. Так алгоритм застрянет в этом действии

Cattharine committed:
Fix requested

Cattharine committed:
Added current sprint hours from context
